commit 5eafc4313ab407d8b8c219c9c5a58115f680eef2
Author: q66 <q66@chimera-linux.org>
Date:   Thu Apr 4 18:24:13 2024 +0200

    improve mdadm startup and make mdadm/dmraid optional
    
    We want to give system startup a shot at very least.

diff --git a/early/scripts/mdadm.sh b/early/scripts/mdadm.sh
index 4faf39f..7eca9f4 100644
--- a/early/scripts/mdadm.sh
+++ b/early/scripts/mdadm.sh
@@ -7,4 +7,19 @@ DINIT_NO_CONTAINER=1
 
 command -v mdadm > /dev/null 2>&1 || exit 0
 
+CONFIG=/etc/mdadm.conf
+ALTCONFIG=/etc/mdadm/mdadm.conf
+
+[ ! -f "$CONFIG" ] && [ -f "$ALTCONFIG" ] && CONFIG="$ALTCONFIG" || :
+
+# no config
+if [ ! -f "$CONFIG" ]; then
+    exit 0
+fi
+
+# no array in config
+if ! grep -q "^ARRAY" "$CONFIG"; then
+    exit 0
+fi
+
 exec mdadm -As
diff --git a/services/early-cryptdisks b/services/early-cryptdisks
index 8129f42..50cd24a 100644
--- a/services/early-cryptdisks
+++ b/services/early-cryptdisks
@@ -5,7 +5,7 @@ command    = ./early/scripts/cryptdisks.sh remaining start
 depends-on = early-devices.target
 depends-on = early-cryptdisks-early
 depends-on = early-dmraid
-depends-on = early-mdadm
 depends-on = early-lvm
 depends-ms = early-root-fsck
+waits-for  = early-mdadm
 options    = starts-on-console
diff --git a/services/early-cryptdisks-early b/services/early-cryptdisks-early
index 8b147cd..2e12b37 100644
--- a/services/early-cryptdisks-early
+++ b/services/early-cryptdisks-early
@@ -4,7 +4,7 @@ type       = scripted
 command    = ./early/scripts/cryptdisks.sh early start
 depends-on = early-devices.target
 depends-on = early-keyboard.target
-depends-on = early-dmraid
-depends-on = early-mdadm
 depends-ms = early-root-fsck
+waits-for  = early-dmraid
+waits-for  = early-mdadm
 options    = starts-on-console
diff --git a/services/early-fs-pre.target b/services/early-fs-pre.target
index bc75334..0b474da 100644
--- a/services/early-fs-pre.target
+++ b/services/early-fs-pre.target
@@ -3,5 +3,5 @@
 type       = internal
 depends-on = early-devices.target
 depends-on = early-cryptdisks
-depends-on = early-dmraid
-depends-on = early-mdadm
+waits-for  = early-dmraid
+waits-for  = early-mdadm
diff --git a/services/early-lvm b/services/early-lvm
index 8047e95..b198919 100644
--- a/services/early-lvm
+++ b/services/early-lvm
@@ -4,6 +4,6 @@ type       = scripted
 command    = ./early/scripts/lvm.sh start
 depends-on = early-devices.target
 depends-on = early-cryptdisks-early
-depends-on = early-dmraid
-depends-on = early-mdadm
 depends-ms = early-root-fsck
+waits-for  = early-dmraid
+waits-for  = early-mdadm
